language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,452 +110,192 @@ impl AbstractState {
+    /// checks if a local may be safely destroyed
+    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
+        match self.locals[&idx] {
+            AbstractValue::Reference(_) => true,
+            AbstractValue::Value(Kind::All) | AbstractValue::Value(Kind::Resource) => false,
+            AbstractValue::Value(Kind::Unrestricted) => !self.is_local_borrowed(idx),
+        }
     }
 
-    /// Return true if local@idx may safely be destroyed
-    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
-        self.local(idx).is_safe_to_destroy()
+    /// checks if the stack frame of the function being analyzed can be safely destroyed.
+    /// safe destruction requires that all references in locals have already been destroyed
+    /// and all values in locals are unrestricted and unborrowed.
+    pub fn is_frame_safe_to_destroy(&self) -> bool {
+        self.locals.values().all(|x| x.is_unrestricted_value())
+            && !self.is_nonce_borrowed(self.frame_root)
     }
 
     /// destroys local@idx
-    /// call only if self.is_local_safe_to_destroy(idx) returns true
     pub fn destroy_local(&mut self, idx: LocalIndex) {
+        checked_precondition!(self.is_local_safe_to_destroy(idx));
         let local = self.locals.remove(&idx).unwrap();
         match local {
-            AbstractValue::Reference(nonce) => self.destroy_nonce(nonce),
-            AbstractValue::Value(kind, borrowed_nonces) => {
-                checked_verify!(kind == Kind::Unrestricted && borrowed_nonces.is_empty());
+            AbstractValue::Reference(nonce) => self.remove_nonce(nonce),
+            AbstractValue::Value(kind) => {
+                checked_verify!(kind == Kind::Unrestricted);
             }
         }
     }
 
-    /// nonce must be fresh
+    /// adds nonce to borrow graph
+    /// nonce must not already be in `self.borrow_graph`
     pub fn add_nonce(&mut self, nonce: Nonce) {
-        self.partition.add_nonce(nonce);
+        self.borrow_graph.add_nonce(nonce);
     }
 
-    /// destroys nonce
-    /// borrows of nonce become borrows of any nonce' such that nonce borrows from nonce'
-    pub fn destroy_nonce(&mut self, nonce: Nonce) {
-        let mut nonce_set = BTreeSet::new();
-        let mut new_locals = BTreeMap::new();
-        let mut new_globals = BTreeMap::new();
-        let mut new_borrows = BTreeMap::new();
-
-        if let Some(borrow_info) = self.borrows.remove(&nonce) {
-            new_borrows = self.strong_propagate(nonce, &borrow_info);
-            match borrow_info {
-                BorrowInfo::BorrowedBy(borrowed_by) => {
-                    nonce_set = borrowed_by.clone();
-                }
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    nonce_set = Self::get_union_of_sets(&BTreeSet::new(), &fields_borrowed_by);
-                }
-            }
-        }
+    /// removes `nonce` from borrow graph
```

Comment:
I have put the postcondition in the remove_nonce of BorrowGraph where the heavy lifting is being performed.  This method is simply a wrapper so I expect it to be "inlined" to avoid need for annotation.